README.md

[![Swifty logo](https://github.com/Oyvindkg/swiftydb/blob/swiftydb-2.0-dev/resources/logo.png)](https://github.com/Oyvindkg/swiftydb/blob/swiftydb-2.0-dev/resources/logo.png)

<br>

A typesafe, pure Swift database offering effortless persistence of objects. It draws inspiration from brilliant libraries such as [ObjectMapper](https://github.com/Hearst-DD/ObjectMapper) and [QueryKit](https://github.com/QueryKit/QueryKit) to create a powerfull and easy-to-use database.

<br>

[![CI Status](https://img.shields.io/travis/Oyvindkg/swiftydb/master.svg?style=flat)](https://travis-ci.org/Oyvindkg/swiftydb)
[![Version](https://img.shields.io/cocoapods/v/SwiftyDB.svg?style=flat)](http://cocoapods.org/pods/SwiftyDB)
[![License](https://img.shields.io/cocoapods/l/SwiftyDB.svg?style=flat)](http://cocoapods.org/pods/SwiftyDB)
[![Platform](https://img.shields.io/cocoapods/p/SwiftyDB.svg?style=flat)](http://cocoapods.org/pods/SwiftyDB)
[![Swift](https://img.shields.io/badge/swift-3-brightgreen.svg?style=flat)](http://cocoapods.org/pods/SwiftyDB)

## Content
[Features](#features)<br />
[Usage](#usage)<br />
&emsp; [Creating the database](#usingTheDatabase)<br />
&emsp;&emsp; [Retrieving objects](#retrievingObjects)<br />
&emsp;&emsp; [Storing objects](#storingObjects)<br />
&emsp;&emsp; [Deleting objects](#deletingObjects)<br />
&emsp; [Defining objects](#definingObjects)<br />
&emsp;&emsp; [Mappable](#mappingObjects)<br />
&emsp;&emsp; [Identifiable](#identifyingObjects)<br />
&emsp;&emsp; [Indexable](#indexingObjects)<br />
&emsp; [Migration](#migratingObjects)<br />
[Installation](#installation)<br />
[Limitations](#limitations)<br />
[Performance](#performance)<br />
[License](#license)

## <a name="features">Main features</a>
- [x] Type safe and developer friendly
- [x] Store nested objects
- [x] Store arrays, sets, and dictionaries
- [x] Asynchronous database access
- [x] Thread safe database operations
- [x] Create custom indices to improve query performance
- [x] Automated migration

## <a name="usage">Usage</a>

### <a name="usingTheDatabase">Creating the database</a>

The easiest way to open a database, is simply providing a name for the database.

```Swift
let database = Database(name: "Westeros")
```

For a more detailed configuration, create a new `Configuration` object, and initiate the database

```Swift
var configuration = Configuration(name: "Westeros")

configuration.directory = "~/some/dir"
configuration.mode      = .sandbox

let database = Database(configuration: configuration)
```

#### <a name="retrievingObjects">Retrieving objects</a>

In order to retrieve all objects of a type from the database, simply call the `get(..)` method with the requested object type as a parameter.
```swift
database.get(Stark.self).then { starks in
  /* Use the retreived starks */
}
```

##### Filtering results

Filtering results is really simple due to SwiftyDB's expressive and powerfull query language  
```Swift
let query = Query.get(Stark.self).where("name" == "Sansa" || "age" < 15)

database.get(using: query).then { starks in
  /* Use the retreived starks */
}
```
It is also possible to filter results based on their nested objects' identifier
```Swift
let lady  = Wolf(name: "Lady")
let ghost = Wolf(name: "Ghost")

let query = Query.get(Stark.self).where("wolf" == lady || "wolf" == ghost)

/* Using a storable object in filters is short hand for `<Storable>.<identifier>` */
database.get(using: query).then { starks in
  /* Use the retreived starks */
}
```

| Operator | Function                                |
|:--------:|:----------------------------------------|
|    ==    | X is equal Y                            |
|    !=    | X is not equal Y                        |
|    <     | X is less than Y                        |
|    >     | X is greater than Y                     |
|    <=    | X is less than or equal to Y            |
|    >=    | X is greater than or equal to Y         |
|    ≈≈    | X matches pattern Y                     |
|    !≈    | X does not match pattern Y              |
|          |                                         |
|    <<    | X is in Y if Y is an array              |
|    !<    | X is not in Y if Y is an array          |
|    <<    | X is the range of Y if Y is a range     |
|    !<    | X is not the range of Y if Y is a range |
|          |                                         |
|    !     | Negation of X                           |
|    &&    | Conjunction of X and Y                  |
|   \|\|   | Disjunction of X and Y                  |


##### Sorting results

Sorting the results is also possible by using the `sort(.., ascending: ..)` method.
```Swift
let query = Query.get(Stark.self).order(by: "name", ascending: true)
```

##### Limitig results

To improve the performance of your queries, SwiftyDB offers simple pagination using the `start(..)` and `max(..)` methods. `start(..)` specifies the index of the first included result element, and `max(..)` specifies the maximum number of elements to be retrieved.
```Swift
/* Retrieve 3 starks, skipping the first 2 results */
let query = Query.get(Stark.self).max(3).skip(2)
```

#### <a name="storingObjects">Storing objects</a>
```Swift
let arya = Stark(name: "Arya", age: 9)

database.add(arya).catch { error in 
  /* Handle any errors */
}
```

#### <a name="deletingObjects">Deleting objects</a>
```Swift
let query = Query.delete(Stark.self).where("name" == "Eddard")

database.delete(using: query).catch { error in
  /* Handle any errors */
}
```

### <a name="definingObjects">Defining objects</a>
Any object can be storable by implementing the `Storable` protocol as an extension. This protocol consists of the `Mappable` and `Identifiable` protocols. This section will show how to make the `Stark` class storable.

```Swift
struct Stark {
  var name: String
  var wolf: Wolf?
  var age: Int

  init(name: String, age: Int) {
    self.name = name
    self.age  = age
  }
}
```

#### <a name="mappingObjects">Mappable</a>
Because the dynamic aspects of the Swift language are limited to read operations, this protocol is used to instantiate new objects and populate them with data dynamically.

```Swift
extension Stark: Mappable {

  /** Create a new object */
  static func mappableObject() -> Mappable {
    return Stark(name: "", age: 0)
  }
  
  /** Used to read and write data to objects */
  mutating func map<M>(using mapper: M) where M : Mapper {
    name <- mapper["name"]
    age  <- mapper["age"]
    wolf <- mapper["wolf"]
  }
}
```

> Inheritance is supported by overriding `map(using: Mapper)` and `mappableObject()` in the subtype. Remember to call `super.mapping(map)`!

#### <a name="identifyingObjects">Identifiable</a>
This protocol is used to identify unique objects in the database. This is necessary for the database to keep track of the individual objects and their references. 

`identifier()` should return the name of a property that uniquely identifies an object.

```Swift
extension Stark: Identifiable {
  
  static func identifier() -> String {
    return "name"
  }
}
```

#### <a name="indexingObjects">Indexable (optional)</a>

Creating an index on frequently queried properties can greatly improve retrieval performance. Create an index by providing a collection of properties to be indexed, and an optional filter used to limit the index domain.

```Swift
extension Stark: Indexable {
  static func indices() -> [AnyIndex] {
    return [
      Index.on("age"),
      Index.on("wolf", "name").where("name" << ["Arya", "Sansa"])
    ]
  }
}
```


### <a name="migratingObjects">Migration</a>

If the database detects that a type does not match the stored data, it will automatically add and remove properties in the database as necessary to reflect the type at all times.

> When changing your models, use `.sandbox` mode in the database `Configuration` to avoid undesired changes in the database. Change it back to `.normal` when you are certain everything is in order. 

## <a name="limitations">Limitations</a>
These are some known limitations of the current version:

- Cannot handle circular references between objects
- It is not possible to query collection properties
- It is not possible to store dictionaries of `Storable` objects

All limitations are ment to be improved as fast as possible. Feel free to contribute 😬

## <a name="performance">Performance</a>

Swifty is created for convenience, not speed. For most applications the performance is more than good enough, but is you are going to add and retrieve thousands of nested objects on a regular basis, this library probably isn't for you. 

Objects with various properties added and retrieved per second

| Hardware | Add with simple properties  | Get with simple properties | Add with collection properties  | Get with collection properties | Add with a nested object | Get with a nested object |
|:---:|:---|:---|:---|:---|:---|:---|
| iPhone 6 | 2500 | 3000 | 1000 | 1000 | 1200 | 1200 |

The performance will be improved in future updates. Until release, the focus will be on functionality

## <a name="installation">Installation</a>

### CocoaPods

SwiftyDB is available through [CocoaPods](http://cocoapods.org). To install
it, simply add the following line to your Podfile:

```ruby
pod "SwiftyDB", "~>2.0"
```

### Carthage
Add this line to your Cartfile
```
github 'Oyvindkg/swiftydb'
```

## Author

Øyvind Grimnes, oyvindkg@yahoo.com

## <a name="License">License</a>

SwiftyDB is available under the MIT license. See the LICENSE file for more info.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
`SwiftyDB` adheres to [Semantic Versioning](http://semver.org/).


## [2.0.0](https://github.com/Oyvindkg/swiftydb/releases/tag/2.0.0) (//2016)
Rebuilt from the ground up to provide a type safe, and versatile database.

**Main features**
* Any structure or class can be made storable in a simple extension
* Store any property type, nested objects, and most collections
   * You can make any property type storable by extending it with `StorableProperty`, which is a `RawRepresentable` with the `RawValue` associated type constrained to either a `String`, `Int`, or `Double` type.
* Create custom indices to improve query performance
* Migrate data when updating your models
   * SwiftyDB automatically updates the database tables as you add or remove properties
* Keep the SwiftyDB integration completely separated from your implementation
   * SwiftyDB is exclusively protocol based, and has no required initializers. This means that you can keep the integration in a separate file to avoid tempering with your project implementation. Replacing SwiftyDB will be as simple as deleting the integration file and removing the pod.

**Supported package managers**
* CocoaPods
* Cathage

### [1.1.3](https://github.com/Oyvindkg/swiftydb/releases/tag/1.1.3) (05/10/2016)

#### Changes
* Removed exclusive tag from example Podfile for Cocoapods 1.0.0

### [1.1.2](https://github.com/Oyvindkg/swiftydb/releases/tag/1.1.2) (05/10/2016)

#### Changes
* Properly escaping column names

### [1.1.1](https://github.com/Oyvindkg/swiftydb/releases/tag/1.1.1) (04/29/2016)

#### Changes
* Fixed breaking bug in when not using primary keys

### [1.1.0](https://github.com/Oyvindkg/swiftydb/releases/tag/1.1.0) (03/31/2016)

#### Changes
* Support for storing NSArrays and NSDictionaries (thanks to @kingslay)
* Now using the new Value protocol instead of SQLiteValue
* Fixes bug related to filtering a property with multiple values

## [1.0.0](https://github.com/Oyvindkg/swiftydb/releases/tag/1.0.0) (01/19/2016)

#### Changes
* Added `lessOrEqual`, and `greaterOrEqual` filters to `Filter`
* New filter syntax for all database queries. `..matchingFilter: ..` instead of `..matchingFitlers: ..`


### [0.9.3](https://github.com/Oyvindkg/swiftydb/releases/tag/0.9.3) (01/19/2016)

#### Changes
* Removed unnecessarily complex query generator
* Performance boost using `TinySQLite ~> 0.3.0`
* `Filter` is no longer a `SequenceType`

### [0.9.2](https://github.com/Oyvindkg/swiftydb/releases/tag/0.9.2) (01/18/2016)

#### Changes
* Now using `TinySQLite ~> 0.2.0`


### [0.9.1](https://github.com/Oyvindkg/swiftydb/releases/tag/0.9.1) (01/18/2016)

#### Changes
* Added `Filter` section to index in README
* Updated features in README


### [0.9.0](https://github.com/Oyvindkg/swiftydb/releases/tag/0.9.0) (01/17/2016)

#### Added
* New `Filter` type used to filter queries

#### Changes
* Use the new `Filter` type to filter queries instead of a `Dictionary`
    * `Filter` conforms to `DictionaryLiteralConvertible`. Therefore it is still possible to define your filters as a `Dictionary`
* Updated query generation
* Delete query now binding values in SQLite


### [0.8.2](https://github.com/Oyvindkg/swiftydb/releases/tag/0.8.2) (01/16/2016)

#### Added
* New documentation generated by Jazzy
  * Located in the 'docs'


### [0.8.1](https://github.com/Oyvindkg/swiftydb/releases/tag/0.8.1) (01/15/2016)

#### Changes
* Removed support for `Float80`
  * It is not supported by iOS
* Query generator classes are internal


### [0.8.0](https://github.com/Oyvindkg/swiftydb/releases/tag/0.8.0) (01/15/2016)

#### Added
* New methods for asynchronous queries to the database
* New `Result` type

#### Changes
* Queries are no longer throwing
* All queries use the new `Result` type for callbacks and return types
  * Allows handling results for sync and async queries to be similar


## [0.7.2](https://github.com/Oyvindkg/swiftydb/releases/tag/0.7.2) (01/12/2016)
Initial release
 


Example/SwiftyDB/Models/Dog.swift
//
//  Dog.swift
//  SwiftyDB
//
//  Created by Øyvind Grimnes on 29/08/16.
//  Copyright © 2016 CocoaPods. All rights reserved.
//

import SwiftyDB

enum Breed: String, StorableProperty {
    case dachs = "dachs"
}

struct Dog: Storable {
    
    var bones           = [Bone(dnr: "12321"), Bone(dnr: "asdasdasdsda"), Bone(), Bone(dnr: "0123"), Bone(dnr: "0123"), Bone(dnr: "0123")]
    var ids: [Double]   = [1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20]
    var dates           = [Date(), Date().addingTimeInterval(123123)]
    var type            = Breed.dachs
    
    var name: String
    var age: Double
    var superBone: Bone?
    var weight: Double
    
    init() {
        superBone = Bone(dnr: "\(arc4random_uniform(100))")
        age       = Double(arc4random_uniform(100))
        name      = "\(arc4random())"
        weight    = 13.3
    }
}

extension Dog: Mappable {
    mutating func map<M>(using mapper: inout M) where M : Mapper {
        name            <- mapper["name"]
        age             <- mapper["age"]
        type            <- mapper["type"]
        weight          <- mapper["weight"]
//        bones           <- mapper["bones"]
        superBone       <- mapper["superBone"]
        dates           <- mapper["dates"]
        ids             <- mapper["numbers"]
    }
    
    static func mappableObject() -> Mappable {
        return Dog()
    }
}

extension Dog: Identifiable {
    static func identifier() -> String {
        return "name"
    }
}

extension Dog: SwiftyDB.Indexable {
    static func indices() -> [SwiftyDB.AnyIndex] {
        return [
            Index.on("age", "weight").where("age" == 5),
            Index.on("name").where("name" != "Jens")
        ]
    }
}
